### ISSUE TITLE
Import local module in script

### ISSUE THREAD

**Issue by Hugovdberg** [snakemake/snakemake#956]
**Is your feature request related to a problem? Please describe.**
The [FAQ](https://snakemake.readthedocs.io/en/stable/project_info/faq.html#i-want-to-import-some-helper-functions-from-another-python-file-is-that-possible) states it should be possible to import a module if it resides in the same directory as the Snakefile. However, it currently only works if it resides in the same folder as the *script* resides in. When organising the scripts into subfolders of the scripts directory this becomes quite a burden because the "common" file should then be copied to each directory. Even a relative import from the parent directory does not work.

**Describe the solution you'd like**
Create a folder that can be imported from any script that's run (for example `workflow/utils`) such that each file can import from `workflow.utils`.

The recommended folder structure would then become something like this:
````
├── .gitignore
├── README.md
├── LICENSE.md
├── workflow
│   ├── rules
|   │   ├── module1.smk
|   │   └── module2.smk
│   ├── envs
|   │   ├── tool1.yaml
|   │   └── tool2.yaml
│   ├── scripts
|   │   ├── script1.py
|   │   └── script2.R
│   ├── notebooks
|   │   ├── notebook1.py.ipynb
|   │   └── notebook2.r.ipynb
│   ├── report
|   │   ├── plot1.rst
|   │   └── plot2.rst
│   ├── utils
|   │   └── __init__.py
│   ├── __init__.py
|   └── Snakefile
├── config
│   ├── config.yaml
│   └── some-sheet.tsv
├── results
└── resources
````

The two `__init__.py` files are added such that when editing from the root directory the `workflow.utils` package can be found.

I currently fix this by adding this snippet to my Snakefile:

````python
onstart:
    import shutil
    from pathlib import Path
    path = Path().absolute()
    if (src_path := path / 'workflow/utils').exists():
        dst_path = path / '.snakemake/scripts/workflow/utils'
        dst_path.mkdir(exist_ok=True, parents=True)
        for file in src_path.rglob("*"):
            dst_file = file.relative_to(src_path)
            if str(dst_file).startswith("."):
                continue
            if file.is_dir():
                (dst_path / dst_file).mkdir(parents=False, exist_ok=True)
            else:
                try:
                    shutil.copy2(src=file, dst=dst_path / dst_file)
                except shutil.SameFileError:
                    pass

        (path / '.snakemake/scripts/workflow/__init__.py').touch()


onsuccess:
    import shutil
    from pathlib import Path
    path = Path().absolute() / '.snakemake/scripts/workflow'
    if path.exists():
        shutil.rmtree(path)
````

I'm not sure how this would work when running on a cluster, I don't have access to a cluster to try this.

**Describe alternatives you've considered**

**Additional context**

**Comment by moritzschaefer** [comment]
bump :)

**Comment by jlumpe** [comment]
I agree that this would be extremely useful. For larger pipelines with a lot of shared code I'm structuring it as a self-contained package that needs to be installed via Pip, but that's overkill in most situations.

**Comment by jlumpe** [comment]
From [this bit of source code](https://github.com/snakemake/snakemake/blob/45cb0821c9e862ff8f6df430f3860f4529404185/snakemake/script/__init__.py#L638-L640), it does look like the directory of the script is added to a list that is then appended to `sys.path` in the preamble of the Snakemake-generated version of the script file. However this is only the case when `PythonScript.generate_preamble()` is called with `is_local=True`. I don't really know anything about Snakemake's internals but that makes it seem like things wouldn't work correctly if running on a cluster.

**Comment by moritzschaefer** [comment]
You mean that I could simply add a utils.py to the scripts/ folder and import it in my script.py? I thought I had tested that (unsuccessfully)

Maybe the cluster-runtime environment already adds the script folder to pythonpath at another point in the coffee...?

**Comment by mmcdermott** [comment]
So I'm far from an expert on Snakemake as I'm just getting started with it, but can you just make your codebase pip installable, including the Snakemake files, then use `importlib.resources` to get access to the root path of the installed version of the package at runtime? This may conflict with the suggested default guidance in terms of repository setup of Snakemake, which doesn't seem as friendly with the typical recommendations for Python packages, but it may be worth a look.

**Comment by gibsramen** [comment]
> can you just make your codebase pip installable, including the Snakemake files, then use `importlib.resources` to get access to the root path of the installed version of the package at runtime?

This is how I've dealt with this problem in the past but it always feels a bit hacky. Would love a cleaner recommended approach.
